diff --git a/core/Makefile b/core/Makefile
index d84152b..60efe00 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -454,6 +454,7 @@ BUILT_RAMDISK_TARGET := $(PRODUCT_OUT)/ramdisk.img
 # We just build this directly to the install location.
 INSTALLED_RAMDISK_TARGET := $(BUILT_RAMDISK_TARGET)
 $(INSTALLED_RAMDISK_TARGET): $(MKBOOTFS) $(INTERNAL_RAMDISK_FILES) | $(MINIGZIP)
+	$(hide) $(shell cp $(LOCAL_PATH)/preinstall.sh $(TARGET_ROOT_OUT))
 	$(call pretty,"Target ram disk: $@")
 	$(hide) $(MKBOOTFS) $(TARGET_ROOT_OUT) | $(MINIGZIP) > $@
 
diff --git a/core/preinstall.sh b/core/preinstall.sh
new file mode 100755
index 0000000..d7642e3
--- /dev/null
+++ b/core/preinstall.sh
@@ -0,0 +1,15 @@
+#!/system/bin/sh
+src=/data/vendor/app/
+target=/data/app/
+if [ ! -d /data/vendor/app ];then
+    return
+fi
+for dir in `ls $src`
+do
+    if [ -n $src/$dir ];then
+        if [ -e $src/$dir/"$dir".apk ];then
+        	mv $src/$dir $target
+		rm -rf $src
+	fi
+    fi
+done
